clippy_lints/src/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,105 @@
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use std::borrow::Borrow;
+    /// use std::hash::{Hash, Hasher};
+    ///
+    /// struct ExampleType {
+    ///     data: String
+    /// }
+    ///
+    /// impl Hash for ExampleType {
+    ///     fn hash<H: Hasher>(&self, state: &mut H) {
+    ///         self.data.hash(state);
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<str> for ExampleType {
+    ///     fn borrow(&self) -> &str {
+    ///         &self.data
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<[u8]> for ExampleType {
+    ///     fn borrow(&self) -> &[u8] {
+    ///         self.data.as_bytes()
+    ///     }
+    /// }
+    /// ```
+    /// As a consequence, hashing a `&ExampleType` and hashing the result of the two
+    /// borrows will result in different values.
+    ///
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SEMANTICS,
+    correctness,
+    "Ensures that the semantics of `Borrow` for `Hash` are satisfied when `Borrow<str>` and `Borrow<[u8]>` are implemented"
+}
+
+declare_lint_pass!(HashBorrowStrSemantics => [HASH_BORROW_STR_SEMANTICS]);
+
+impl LateLintPass<'_> for HashBorrowStrSemantics {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of `Hash`, `Borrow<str>` and `Borrow<[u8]>`.
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if let ItemKind::Impl(imp) = item.kind
+            && let Some(TraitRef {path: Path {span, res, ..}, ..}) = imp.of_trait
+            && let ty = cx.tcx.type_of(item.owner_id).instantiate_identity()
+            && let Some(hash_id) = cx.tcx.get_diagnostic_item(sym::Hash)
+            && Res::Def(DefKind::Trait, hash_id) == *res
+            && let Some(borrow_id) = cx.tcx.get_diagnostic_item(sym::Borrow)
+            // since we are in the `Hash` impl, we don't need to check for that.
+            // we need only to check for `Borrow<str>` and `Borrow<[u8]>`
+            && implements_trait(cx, ty, borrow_id, &[cx.tcx.types.str_.into()])
+            && implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()])
+        {
+            span_lint_and_then(
+                cx,
+                HASH_BORROW_STR_SEMANTICS,
+                *span,
+                "can't satisfy the semantics of `Hash` when both `Borrow<str>` and `Borrow<[u8]>` are implemented",
+                |diag| {
+                    diag.note(
+                        "`Borrow` requires `Hash` values for all borrowed types to be the same. This is not the case for `Borrow<str>` and `Borrow<[u8]>`"
+                    );
+                    diag.note("as the two types result in different `Hash` values, e.g. `hash(\"abc\") != hash(\"abc\".as_bytes())`");
+
+                    diag.help("consider removing one implementation of `Borrow` (`Borrow<str>` or `Borrow<[u8]>`) or not implementing `Hash` for this type");
+                },
+            );
```

Comment:
Creating a good message for this lint is hard, due to the complex nature of the issue. I have the feeling, that the following code snippet will make the issue a bit clearer and is more inline to the [message guidelines](https://rustc-dev-guide.rust-lang.org/diagnostics.html#error-messages), but I can't really define what makes it better :sweat_smile: 

```rs
span_lint_and_then(
    cx,
    HASH_BORROW_STR_SEMANTICS,
    *span,
    "implementation of `Hash` that doesn't satisfy the semantics of the trait due to other impls",
    |diag| {
        diag.note("the type of this implementation implements `Borrow<str>` and `Borrow<[u8]>`");
        diag.note("the `Borrow` semantics define that the `Hash` values for all borrowed types should be the same");
        diag.note(
            "however, a string and the bytes of a string have different hash values (`hash(\"abc\") != hash(\"abc\".as_bytes())`)."
        );

        diag.help("consider either removing one of the  `Borrow` implementations (`Borrow<str>` or `Borrow<[u8]>`) ...");
        diag.help("... or don't implement `Hash` for this type");
    },
);
```

The resulting message would roughly look like this:

```
error: implementation of `Hash` that doesn't satisfy the semantics of the trait due to other impls
  --> $DIR/hash_borrow_str_semantics.rs:10:6
   |
LL | impl Hash for ExampleType {
   |      ^^^^
   |
   = note: the type of this implementation implements `Borrow<str>` and `Borrow<[u8]>`
   = note: the `Borrow` trait define that the `Hash` values for all borrowed types should be the same
   = note: however, a string and the bytes of a string have different hash values (`hash(\"abc\") != hash(\"abc\".as_bytes())`)
   = help: consider either removing one of the  `Borrow` implementations (`Borrow<str>` or `Borrow<[u8]>`) ...
   = help: ... or don't implement `Hash` for this type
```

Reply:
I think it's the wording of the semantic s, in particular the whole "`Hash` values" sounds a bit wrong.

Reply:
You're right, is "hashes" or something better?

Reply:
I am not sure, current lint is

```rust
 "the semantics of `Borrow<T>` around `Hash` can't be satisfied when both `Borrow<str>` and `Borrow<[u8]>` are implemented",
  |diag| {
      diag.note("the `Borrow` semantics define that the `Hash` values for all borrowed types should be the same");
      diag.note(
          "however, a string and the bytes of a string have different hash values (`hash(\"abc\") != hash(\"abc\".as_bytes())`)"
      );
      diag.help("consider either removing one of the  `Borrow` implementations (`Borrow<str>` or `Borrow<[u8]>`) ...");
      diag.help("... or don't implement `Hash` for this type");
  },
```

perhaps this is more appropriate?

```rust
 "the semantics of `Borrow<T>` around `Hash` can't be satisfied when both `Borrow<str>` and `Borrow<[u8]>` are implemented",
  |diag| {
      diag.note("the `Borrow` semantics define that `Hash` must behave the same for all implementations of Borrow<T>");
      diag.note(
          "however, the hash implementations of a string (`str`) and the bytes of a string `[u8]` do not behave the same"
      );
      diag.note("as (`hash(\"abc\") != hash(\"abc\".as_bytes())`");
      diag.help("consider either removing one of the  `Borrow` implementations (`Borrow<str>` or `Borrow<[u8]>`) ...");
      diag.help("... or don't implement `Hash` for this type");
  },
```

Reply:
Yeah, I think having it like the second example would be good. Then everything should be ready :)

Reply:
Great! :D I adjusted the last sentence a bit as it was grammatically incorrect 😓 given the changes.